### PR TITLE
Issue758

### PR DESCRIPTION
**Pull request by aklish**
- For JSON-API, the preCommit life cycle hooks will now run prior to response serialization.
- This work has not been done for GraphQL.  That is a more complex change given the lack of control over the graphQL parser.
- Added a new mechanism to add dependency injection to Elide.  Ultimately, the hope is that the same mechanism can be used to inject checks (as well as models).
- Made IT tests support dependency injection
- Added a new package of example/test models for testing lifecycle hooks.  This has a canonical example of how to inject a service and invoke on an lifecycle hook.

### CODE REVIEW

File: elide-core/src/main/java/com/yahoo/elide/core/EntityDictionary.java [hunk 1]
```diff
@@ -694,6 +712,8 @@ public String getNameFromAlias(Object entity, String alias) {
             Initializer<T> initializer = getEntityBinding(entity.getClass()).getInitializer();
             if (initializer != null) {
                 initializer.initialize(entity);
+            } else if (injector != null) {
```
Comment (wcekan) [inline elide-core/src/main/java/com/yahoo/elide/core/EntityDictionary.java]: @aklish what do we expect if both are defined?

File: elide-example-models/src/main/java/com/yahoo/elide/models/triggers/Invoice.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    @Transient
+    @Exclude
+    @Inject
+    private BillingService billingService;
+
+    private boolean complete = false;
+
+    private long total = 0;
+
+
```
Comment (jkusa) [inline elide-example-models/src/main/java/com/yahoo/elide/models/triggers/Invoice.java]: `\n`

File: elide-core/src/main/java/com/yahoo/elide/core/EntityDictionary.java (hunk 1 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/EntityDictionary.java]: The initializer overrides the injector.  There is only one way to initialize something - either a general injector or a class specific initializer.

File: elide-core/src/main/java/com/yahoo/elide/core/EntityDictionary.java [hunk 3]
```diff
@@ -80,13 +82,29 @@
...
+     * Instantiate a new EntityDictionary with the provided set of checks and an injection function.
+     * In addition all of the checks * in {@link com.yahoo.elide.security.checks.prefab} are mapped
+     * to {@code Prefab.CONTAINER.CHECK} * (e.g. {@code @ReadPermission(expression="Prefab.Role.All")}
+     * or {@code @ReadPermission(expression="Prefab.Common.UpdateOnCreate")})
+     * @param checks a map that links the identifiers used in the permission expression strings
+     *               to their implementing classes
+     * @param injector a function typically associated with a dependency injection framework that will
+     *                 initialize Elide models.
+     */
+    public EntityDictionary(Map<String, Class<? extends Check>> checks, Initializer injector) {
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/core/EntityDictionary.java]: Can we make this `Injector injector` to make the distinction clear.

**Pull request merged**